File: tours_and_travels/tours_and_travels/report/restaurant_wise_contract/restaurant_wise_contract.py
# Copyright (c) 2022, Nihantra C. Patel and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
	columns, data = get_columns(filters), get_data(filters)
	return columns, data

def get_columns(filters):
	columns = [
			{
				"label": _("Contract Name"),
				"fieldname": "name",
				"fieldtype": "Link",
				"options": "Contract",
				"width": 120
			},
			{
				"label": _("Party Name"),
				"fieldname": "party_name",
				"fieldtype": "Link",
				"options": "Supplier",
				"width": 120
			},
			{
				"label": _("Travelling Seasons"),
				"fieldname": "travelling_seasons",
				"fieldtype": "Link",
				"options": "Travelling Seasons",
				"width": 140
			},
			{
				"label": _("Start Date"),
				"fieldname": "start_date",
				"fieldtype": "Date",
				"width": 100
			},
			{
				"label": _("End Date"),
				"fieldname": "end_date",
				"fieldtype": "Date",
				"width": 100
			},
			{
				"label": _("Restaurant"),
				"fieldname": "hotel_restaurant",
				"fieldtype": "Link",
				"options": "Restaurant",
				"width": 200
			},
			{
				"label": _("Meal Type"),
				"fieldname": "meal_type",
				"fieldtype": "Link",
				"options": "Meal Type",
				"width": 120
			},
			{
				"label": _("Qty"),
				"fieldname": "qty",
				"fieldtype": "Int",
				"width": 90
			},
			{
				"label": _("Unit Price"),
				"fieldname": "unit_price",
				"fieldtype": "Currency",
				"width": 100
			},
			{
				"label": _("Signed"),
				"fieldname": "is_signed",
				"fieldtype": "Check",
				"width": 70
			},
			{
				"label": _("Signee"),
				"fieldname": "signee",
				"fieldtype": "Data",
				"width": 100
			},
			{
				"label": _("Signed On"),
				"fieldname": "signed_on",
				"fieldtype": "Datetime",
				"width": 130
			},
			{
				"label": _("City"),
				"fieldname": "city",
				"fieldtype": "Link",
				"options": "City",
				"width": 100
			},
			{
				"label": _("State"),
				"fieldname": "state",
				"fieldtype": "Link",
				"options": "State",
				"width": 100
			},
			{
				"label": _("Country"),
				"fieldname": "country",
				"fieldtype": "Link",
				"options": "Country",
				"width": 100
			}
		]
	return columns

def get_data(filters):
	return frappe.db.sql(
			"""
			SELECT
				ctr.name, ctr.party_name, ctr.travelling_seasons, ctr.start_date, ctr.end_date, ctr.hotel_restaurant,
				mpl.meal_type, mpl.qty, mpl.unit_price, ctr.is_signed,
				ctr.signee, ctr.signed_on, ctr.city, ctr.state, ctr.country

			FROM
				`tabContract` as ctr, `tabMeal Package List` as mpl

			WHERE
				mpl.parent = ctr.name
				AND ctr.docstatus < 2
				AND ctr.package_contract_type = "Restaurant"
				AND ctr.start_date BETWEEN %(start_from_date)s AND %(start_to_date)s
				AND ctr.end_date BETWEEN %(end_from_date)s AND %(end_to_date)s

				{conditions}

			ORDER BY
				ctr.end_date DESC """.format(
				conditions=get_conditions(filters)
			),
			filters,
			as_dict=1,
		)

def get_conditions(filters):
	conditions = []

	if filters.get("name"):
		conditions.append(" AND ctr.name in %(name)s")

	if filters.get("party_name"):
		conditions.append(" AND ctr.party_name in %(party_name)s")

	if filters.get("travelling_seasons"):
		conditions.append(" AND ctr.travelling_seasons in %(travelling_seasons)s")

	if filters.get("hotel_restaurant"):
		conditions.append(" AND ctr.hotel_restaurant in %(hotel_restaurant)s")

	if filters.get("meal_type"):
		conditions.append(" AND mpl.meal_type in %(meal_type)s")

	if filters.get("city"):
		conditions.append(" AND ctr.city in %(city)s")

	if filters.get("state"):
		conditions.append(" AND ctr.state in %(state)s")

	if filters.get("country"):
		conditions.append(" AND ctr.country in %(country)s")

	return " ".join(conditions) if conditions else ""

File: tours_and_travels/tours_and_travels/report/restaurant_wise_contract/restaurant_wise_contract.js
// Copyright (c) 2022, Nihantra C. Patel and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Restaurant-wise Contract"] = {
    "filters": [
        {
            "fieldname": "start_from_date",
            "label": __("Start From Date"),
            "fieldtype": "Date",
            "default": frappe.defaults.get_user_default("year_start_date"),
            "reqd": 1
        },
        {
            "fieldname": "start_to_date",
            "label": __("Start To Date"),
            "fieldtype": "Date",
            "default": frappe.defaults.get_user_default("year_end_date"),
            "reqd": 1
        },
        {
            "fieldname": "end_from_date",
            "label": __("End From Date"),
            "fieldtype": "Date",
            "default": frappe.defaults.get_user_default("year_start_date"),
            "reqd": 1
        },
        {
            "fieldname": "end_to_date",
            "label": __("End To Date"),
            "fieldtype": "Date",
            "default": frappe.defaults.get_user_default("year_end_date"),
            "reqd": 1
        },
        {
            "fieldname": "name",
            "label": __("Contract Name"),
            "fieldtype": "MultiSelectList",
            "options": "Contract",
            get_data: function(txt) {
                return frappe.db.get_link_options('Contract', txt, {
                	package_contract_type: "Restaurant"
                });
            }
        },
        {
            "fieldname": "party_name",
            "label": __("Party Name"),
            "fieldtype": "MultiSelectList",
            "options": "Supplier",
            get_data: function(txt) {
                return frappe.db.get_link_options('Supplier', txt)
            }
        },
        {
            "fieldname": "travelling_seasons",
            "label": __("Travelling Seasons"),
            "fieldtype": "MultiSelectList",
            "options": "Travelling Seasons",
            get_data: function(txt) {
                return frappe.db.get_link_options('Travelling Seasons', txt);
            }
        },
        {
            "fieldname": "hotel_restaurant",
            "label": __("Restaurant"),
            "fieldtype": "MultiSelectList",
            "options": "Restaurant",
            get_data: function(txt) {
                return frappe.db.get_link_options('Restaurant', txt);
            }
        },
        {
            "fieldname": "meal_type",
            "label": __("Meal Type"),
            "fieldtype": "MultiSelectList",
            "options": "Meal Type",
            get_data: function(txt) {
                return frappe.db.get_link_options('Meal Type', txt);
            }
        },
        {
            "fieldname": "city",
            "label": __("City"),
            "fieldtype": "MultiSelectList",
            "options": "City",
            get_data: function(txt) {
                return frappe.db.get_link_options('City', txt);
            }
        },
        {
            "fieldname": "state",
            "label": __("State"),
            "fieldtype": "MultiSelectList",
            "options": "State",
            get_data: function(txt) {
                return frappe.db.get_link_options('State', txt);
            }
        },
        {
            "fieldname": "country",
            "label": __("Country"),
            "fieldtype": "MultiSelectList",
            "options": "Country",
            get_data: function(txt) {
                return frappe.db.get_link_options('Country', txt);
            }
        }
    ],

    onload: function(report) {
        report.page.set_inner_btn_group_as_primary(__('Contract Reports'));

        report.page.add_inner_button(("Hotel-wise Contract"), function() {
            var filters = report.get_values();
            frappe.set_route('query-report', 'Hotel-wise Contract', {
                start_from_date: filters.start_from_date,
                start_to_date: filters.start_to_date,
                dep_from_date: filters.dep_from_date,
                dep_to_date: filters.dep_to_date,
                name: filters.name,
                party_name: filters.party_name,
                travelling_seasons: filters.travelling_seasons,
                city: filters.city,
                state: filters.state,
                country: filters.country
            });
        }, __('Contract Reports'));

        report.page.add_inner_button(("Transportation-wise Contract"), function() {
            var filters = report.get_values();
            frappe.set_route('query-report', 'Transportation-wise Contract', {
                start_from_date: filters.start_from_date,
                start_to_date: filters.start_to_date,
                dep_from_date: filters.dep_from_date,
                dep_to_date: filters.dep_to_date,
                name: filters.name,
                party_name: filters.party_name,
                travelling_seasons: filters.travelling_seasons,
                city: filters.city,
                state: filters.state,
                country: filters.country
            });
        }, __('Contract Reports'));

    }
};